---
title: "ex4"
author: "yuval aharon"
date: "21 11 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
install.packages("gProfileR")
install.packages("knitr")
BiocManager::install("gage")
```

```{r}
library(compGenomRData)
library(DESeq2)
library(gProfileR)
library(gage)
```

```{r}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
                           package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
                            package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))

#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))

#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
                      stringsAsFactors = TRUE)

#define the design formula
designFormula <- "~ group"

#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = as.formula(designFormula))
dds <- DESeq(dds)

DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))
```

```{r message = F, warning=FALSE}
#remove genes with NA values
DE <- DEresults[!is.na(DEresults$padj),]

#select genes with adjusted p-values below 0.1
DE <- DE[DE$padj < 0.1,]

#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]

#get the list of genes of interest
genesOfInterest <- rownames(DE)

#calculate enriched GO terms
goResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate')

goResults = goResults[order(goResults$p.value),]
goResults = goResults[order(goResults$precision),]
goResults = goResults[order(goResults$recall),]
```

1.
```{r}
goResults2 <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'KEGG',
                       hier_filtering = 'moderate')

# Sort the resulting tables by columns precision and/or recall. How do the top GO terms change when sorted for precision, recall, or p.value? hint: use order() for sorting

goResults2 = goResults2[order(goResults2$p.value),]
goResults2 = goResults2[order(goResults2$precision),]
goResults2 = goResults[order(goResults$recall),]
```

```{r}
goResults3 <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'REACTOME',
                       hier_filtering = 'moderate')

# Sort the resulting tables by columns precision and/or recall. How do the top GO terms change when sorted for precision, recall, or p.value? hint: use order() for sorting

goResults3 = goResults3[order(goResults3$p.value),]
goResults3 = goResults3[order(goResults3$precision),]
goResults3 = goResults[order(goResults$recall),]
```

```{r}
goResults4 <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'CORUM',
                       hier_filtering = 'moderate')

# Sort the resulting tables by columns precision and/or recall. How do the top GO terms change when sorted for precision, recall, or p.value? hint: use order() for sorting

goResults4 = goResults4[order(goResults4$p.value),]
goResults4 = goResults4[order(goResults4$precision),]
goResults4 = goResults[order(goResults$recall),]
```

2.

```{r}

#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue

goResults <- goResults[order(goResults$p.value),]

#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]

# use the top term from this table to create a gene set
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))

#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)
geneSet2 <- sample(rownames(normalizedCounts), 25)
geneSets <- list('top_GO_term' = geneSet1,
'random_set' = geneSet2)

# Using the defined gene sets, weâ€™d like to do a group comparison between the case
# samples with respect to the control samples.
#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')
summary(gseaResults)

gseaResults2 = gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'paired')
summary(gseaResults2)
```

3.
```{r}
goResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'CORUM',
                       hier_filtering = 'moderate',
                       significant = F)
plot(goResults$term.size,-log10(goResults$p.value))


```

4.
```{r Do a gene-set enrichment analysis using gene sets from top 10 GO terms}

if (!requireNamespace("BiocManager", quietly=TRUE))
+ install.packages("BiocManager")
BiocManager::install("topGO")


topgeneSet1 <- unlist(strsplit(go[1,]$intersection, ','))
topgeneSet2 <- unlist(strsplit(go[2,]$intersection, ','))
topgeneSet3 <- unlist(strsplit(go[3,]$intersection, ','))
topgeneSet4 <- unlist(strsplit(go[4,]$intersection, ','))
topgeneSet5 <- unlist(strsplit(go[5,]$intersection, ','))
topgeneSet6 <- unlist(strsplit(go[6,]$intersection, ','))
topgeneSet7 <- unlist(strsplit(go[7,]$intersection, ','))
topgeneSet8 <- unlist(strsplit(go[8,]$intersection, ','))
topgeneSet9 <- unlist(strsplit(go[9,]$intersection, ','))
topgeneSet10 <- unlist(strsplit(go[10,]$intersection, ','))

allgeneSets <- list('GO_term1' = topgeneSet1, 'GO_term2' = topgeneSet2, 'GO_term3' = topgeneSet3, 'GO_term4' = topgeneSet4, 'GO_term5' = topgeneSet5, 'GO_term6' = topgeneSet6, 'GO_term7' = topgeneSet7,'GO_term8' = topgeneSet8, 'GO_term9' = topgeneSet9, 'GO_term10' = topgeneSet10,'random_set' = geneSet2)

gseaResults2 <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = allgeneSets, compare = 'as.group')

summary(gseaResults2)

```

5.
among the packages i found: EGSEA, topGO, fgsea, MAVTgsa and RGSEA.

6.

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("topGO")
```



```{r}
library(topGO)
BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
pvalforGO = DE[genesOfInterest,"pvalue"]
names(pvalforGO) = genesOfInterest

# Create topGOData object

selection <- function(allScore){ return(allScore < 0.05)} # function that returns TRUE/FALSE for p-values<0.05
allGO2genes <- annFUN.org(whichOnto="BP", feasibleGenes=NULL, mapping="org.Hs.eg.db", ID="symbol")
GOdata <- new("topGOdata",
  ontology="BP",
  allGenes=pvalforGO,
  annot=annFUN.GO2genes,
  GO2genes=allGO2genes,
  geneSel=selection,
  nodeSize=10)

#In order to make use of the rank information, use Kolmogorov-Smirnov (K-S) test:
results.ks <- runTest(GOdata, algorithm="classic", statistic="ks")
goEnrichment <- GenTable(GOdata, KS=results.ks, orderBy="KS", topNodes=20)
goEnrichment$KS <- as.numeric(goEnrichment$KS)
goEnrichment <- goEnrichment[goEnrichment$KS<0.05,]
goEnrichment <- goEnrichment[,c("GO.ID","Term","KS")]
goEnrichment$Term <- gsub(" [a-z]*\\.\\.\\.$", "", goEnrichment$Term)
goEnrichment$Term <- gsub("\\.\\.\\.$", "", goEnrichment$Term)
goEnrichment$Term <- paste(goEnrichment$GO.ID, goEnrichment$Term, sep=", ")
goEnrichment$Term <- factor(goEnrichment$Term, levels=rev(goEnrichment$Term))

summary(goEnrichment)
```

7.
```{r}
BiocManager::install("biomaRt")
library(biomaRt)
Sys.setenv("http_proxy" = "http://my.proxy.org:9999")
options(RCurlOptions = list(proxy="uscache.kcc.com:80",proxyuserpwd="------:-------"))

human = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl")
C.elegans = biomaRt::useMart("ensembl", dataset = "celegans_gene_ensembl")
biomaRt::getLDS(attributes = c("hgnc_symbol","chromosome_name","start_position"),
       filters = "hgnc_symbol", values = "TP53",mart = human,
      attributesL = c("refseq_peptide","chromosome_name","start_position"), martL = C.elegans)
```

8.

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("org.Mm.eg.db")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("gseKEGG")

BiocManager::install("gseKEGG")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("gage")

```

```{r}
install.packages("msigdbr")
library(msigdbr)
msigdbr_df <- msigdbr(species = "human", category = "C2")
msigdbrEntrez = as.list(msigdbr_df$entrez_gene)

gseaResults3 <-gage::gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = msigdbrEntrez, compare = 'as.group')
summary(gseaResults3)
```

```{r}



```
